src/response/chi0.jl

```diff
@@ -239,94 +239,119 @@ function compute_αmn(fm, fn, ratio)
+        # First compute δocc without self-consistent Fermi δεF.
         D = zero(T)
-        for ik = 1:Nk, (n, εnk) in enumerate(ε_occ[ik])
+        for ik = 1:Nk, (n, εnk) in enumerate(ε[ik])
             enred = (εnk - εF) / temperature
-            δεnk = real(dot(ψ_occ[ik][:, n], δHψ[ik][:, n]))
+            δεnk = real(dot(ψ[ik][:, n], δHψ[ik][:, n]))
             fpnk = filled_occ * Smearing.occupation_derivative(smearing, enred) / temperature
             δocc[ik][n] = δεnk * fpnk
             D += fpnk * basis.kweights[ik]
         end
-        # compute δεF
+        # Compute δεF…
         D = mpi_sum(D, basis.comm_kpts)  # equal to minus the total DOS
         δocc_tot = mpi_sum(sum(basis.kweights .* sum.(δocc)), basis.comm_kpts)
         δεF = !isnothing(model.εF) ? zero(δεF) : δocc_tot / D  # no δεF when Fermi level is fixed
-        # recompute δocc
-        for ik = 1:Nk, (n, εnk) in enumerate(ε_occ[ik])
+        # … and recompute δocc, taking into account δεF.
+        for ik = 1:Nk, (n, εnk) in enumerate(ε[ik])
             enred = (εnk - εF) / temperature
             fpnk = filled_occ * Smearing.occupation_derivative(smearing, enred) / temperature
             δocc[ik][n] -= fpnk * δεF
         end
     end
 
-    # compute δψnk band per band
-    δψ = zero.(ψ)
+    (; δocc, δεF)
+end
+
+"""
+Compute the derivatives of the wave functions by solving a Sternheimer equation for each
+`k`-points.
+"""
+function compute_δψ(basis, H, ψ, εF, ε, δHψ; ψ_extra=[zeros(size(ψk,1), 0) for ψk in ψ],
+                    kwargs_sternheimer...)
+    model = basis.model
+    temperature = model.temperature
+    smearing = model.smearing
+    filled_occ = filled_occupation(model)
+    Nk = length(basis.kpoints)
+
+    # Compute δψnk band per band
+    δψ = zero.(δHψ)
     for ik = 1:Nk
-        ψk = ψ_occ[ik]
+        Hk  = H[ik]
+        ψk  = ψ[ik]
+        εk  = ε[ik]
         δψk = δψ[ik]
-        Hψk_extra = ham.blocks[ik] * ψ_extra[ik]
 
-        εk = ε_occ[ik]
+        ψk_extra = ψ_extra[ik]
+        Hψk_extra = Hk * ψk_extra
+        εk_extra  = diag(real.(ψk_extra' * Hψk_extra))
         for n = 1:length(εk)
             fnk = filled_occ * Smearing.occupation(smearing, (εk[n]-εF) / temperature)
 
-            # explicit contributions (nonzero only for temperature > 0)
+            # Explicit contributions (nonzero only for temperature > 0)
             for m = 1:length(εk)
+                εk[m] == εk[n] && continue
                 fmk = filled_occ * Smearing.occupation(smearing, (εk[m]-εF) / temperature)
                 ddiff = Smearing.occupation_divided_difference
                 ratio = filled_occ * ddiff(smearing, εk[m], εk[n], εF, temperature)
                 αmn = compute_αmn(fmk, fnk, ratio)  # fnk * αmn + fmk * αnm = ratio
-                δψk[:, n] .+= ψk[:, m] .* αmn .* (dot(ψk[:, m], δHψ[ik][:, n]) * (n != m))
+                δψk[:, n] .+= ψk[:, m] .* αmn .* dot(ψk[:, m], δHψ[ik][:, n])
```

Comment:
why?

Reply:
@gkemlin I don't remember why this was, but this commit still passes tests, so possibly our test suite is too coarse?

Reply:
Hum, that's weird. On master, if I remove `n != m`, the `chi0.jl` test fails as expected. This is because the `n==m` term is taken into account when computing `δρ` through `δoccupation`. Are you sure you didn't forget to specifiy `δoccupation` in `compute_δρ` later on ?

Reply:
OK great, j'ai rien dit!

Reply:
Can't we have `εk[m] == εk[n]` even though `m!=n`. In theory yes, but in practice hopefully not because of the tolerance on eigensolvers, but I'm worried that this would fail with bad luck ? Also, the condition `m != n` seems more explicit to me.

Reply:
Ah yes! I merged too early... Will make a followup commit